fabric-enchantments-v1/src/main/java/net/fabricmc/fabric/mixin/enchantment/MixinEnchantingTableContainer.java

```diff
@@ -0,0 +1,67 @@
+/*
+ * Copyright (c) 2016, 2017, 2018, 2019 FabricMC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package net.fabricmc.fabric.mixin.enchantment;
+
+import net.fabricmc.fabric.api.enchantment.EnchantingPowerProvider;
+import net.minecraft.block.BlockState;
+import net.minecraft.container.EnchantingTableContainer;
+import net.minecraft.item.ItemStack;
+import net.minecraft.util.math.BlockPos;
+import net.minecraft.world.World;
+import org.spongepowered.asm.mixin.Mixin;
+import org.spongepowered.asm.mixin.injection.At;
+import org.spongepowered.asm.mixin.injection.Inject;
+import org.spongepowered.asm.mixin.injection.ModifyVariable;
+import org.spongepowered.asm.mixin.injection.callback.CallbackInfo;
+
+@Mixin(EnchantingTableContainer.class)
+public abstract class MixinEnchantingTableContainer {
+	private World fabric_world;
+	private BlockPos fabric_blockPos;
+
+	@Inject(method = "method_17411", at = @At("HEAD"))
+	private void onEnchantmentCalculation(ItemStack itemStack, World world, BlockPos blockPos, CallbackInfo callbackInfo) {
+		fabric_world = world;
+		fabric_blockPos = blockPos;
+	}
+
+	@ModifyVariable(method = "method_17411", at = @At(value = "INVOKE", target = "Ljava/util/Random;setSeed(J)V", shift = At.Shift.BEFORE), ordinal = 0)
+	private int changeEnchantingPower(int power) {
+		for(int zOffset = -1; zOffset <= 1; ++zOffset) {
+			for(int xOffset = -1; xOffset <= 1; ++xOffset) {
+				if ((zOffset != 0 || xOffset != 0) && fabric_world.isAir(fabric_blockPos.add(xOffset, 0, zOffset)) && fabric_world.isAir(fabric_blockPos.add(xOffset, 1, zOffset))) {
```

Comment:
Would this benefit from using a mutable BlockPos ? And is there no way to inject directly into the vanilla loop instead ?

Reply:
> Would this benefit from using a mutable BlockPos ?

Will a mutable BlockPos be more efficient? Idk much about this class.
> And is there no way to inject directly into the vanilla loop instead ?

That wouldn't be too hard. Is it better to have an injection called *n* times or to have it have its own loop?

Reply:
Mutable BlockPos is more efficient when you iterate over a lot of positions, as it avoids allocating a bunch of throwaway objects. 
~~In this case, I think having the injection being called n times would be more efficient, but I am not sure because of the CallbackInfo object.~~ You are using a `ModifyVariable` injection, so hell yes it would be more efficient (you only add an INVOKEVIRTUAL per loop). The main reason to do that however is to avoid copying vanilla code which may change subtly over versions. And as a rule of thumb, you should always minimize the logic in mixins.

Reply:
(addendum: since the method is private, it will be an INVOKESPECIAL, which has an even more negligible performance cost)

Reply:
Okay, I tried to implement it in the loop. I got a problem now:
Two make it work in the loop I need to capture the local loop indexes. Therefore I would need to inject **and** modify the variable each time.

I guess I'll take the first approach and change it to use a `BlockPos.Mutable` and let it be it.

Reply:
Ok, I implemented the use of mutable BlockPos. Is it better now?

Reply:
Mixin is a bit annoying sometimes... But I still think an injection inside the loop is preferable. It reduces the amount of logic in the mixin, it minimizes the changes in case of vanilla updates, it maximizes compatibility with other mods (imagine a mod that makes the scan radius wider), and it avoids doing twice the world lookups.